controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -507,32 +514,283 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
+                         .thenCompose(v -> {
+                             StreamCutRecord limitMax = limitMaxFuture.join();
+                             StreamCutRecord limitMin = limitMinFuture.join();
+                             if (limitMin != null) {
+                                 return streamMetadataStore.streamCutStrictlyGreaterThan(scope, stream, limitMin.getStreamCut(), lowerBound, context, executor)
+                                                    .thenCompose(gtMin -> {
+                                                        if (gtMin) {
+                                                            if (limitMax == null) {
+                                                                return CompletableFuture.completedFuture(lowerBound);
+                                                            } else {
+                                                                return streamMetadataStore.streamCutStrictlyGreaterThan(scope, stream, limitMax.getStreamCut(), lowerBound, context, executor)
+                                                                                          .thenApply(gtMax -> gtMax ? limitMax.getStreamCut() : lowerBound);
+                                                            }
+                                                        } else {
+                                                            return streamMetadataStore.streamCutStrictlyGreaterThan(scope, stream, lowerBound, limitMin.getStreamCut(), context, executor)
+                                                                               .thenApply(gt -> gt ? limitMin.getStreamCut() : null);
+                                                        }
+                                                    });
+                             } else {
+                                return CompletableFuture.completedFuture(null);  
+                             }
+                         });
+    }
+
+    private Map<Long, Long> computeSubscribersLowerBound(List<Map<StreamSegmentRecord, Long>> subscribers) {
+        // loop over all streamcuts and for each segment in new streamcut: 
+        // if new segment is predecessor of any segment in the bound then replace the successor segment with
+        // this segment. 
+        Map<StreamSegmentRecord, Long> lowerBound = new HashMap<>();
+        subscribers.forEach(streamCut -> streamCut.forEach((segment, offset) -> {
+            if (lowerBound.containsKey(segment)) {
+                if (lowerBound.get(segment) > offset) {
+                    lowerBound.put(segment, offset);
+                }
+            } else {
+                Map<StreamSegmentRecord, Long> predecessors = new HashMap<>();
+                Map<StreamSegmentRecord, Long> successors = new HashMap<>();
+                lowerBound.forEach((s, o) -> {
+                    if (s.overlaps(segment)) {
+                        if (s.segmentId() < segment.segmentId()) {
```

Comment:
Q: We can use the getCreationEpoch() here right?

Reply:
creation epoch is extracted from the segment id itself. we create segmentid as "32byte epoch + 32 byte segment number".